File: scripts/tasks/verify-config-backup.ts
import { createDecipheriv, pbkdf2Sync } from "node:crypto";
import {
  existsSync,
  mkdtempSync,
  readFileSync,
  rmSync,
  statSync,
  writeFileSync,
} from "node:fs";
import { homedir, tmpdir } from "node:os";
import { join, resolve } from "node:path";
import { logStep, runOrThrow } from "../common";

const CONFIG_KDF_ITERATIONS = 250_000;
const CONFIG_KEY_LENGTH = 32;
const CONFIG_SALT_LENGTH = 16;
const CONFIG_IV_LENGTH = 12;
const CONFIG_AUTH_TAG_LENGTH = 16;
const CONFIG_FORMAT_MAGIC = "SCRYCFG1";

const REQUIRED_RESTORE_PATHS = [
  ".openclaw/openclaw.json",
  ".openclaw/credentials",
  ".openclaw/cron/jobs.json",
  ".openclaw/identity",
] as const;

export function verifyConfigBackup(): void {
  const passphrase = process.env.SCRY_CONFIG_BACKUP_PASSPHRASE ?? "";
  const repoRoot = resolve(".");
  const encryptedFile =
    process.env.SCRY_CONFIG_BACKUP_FILE ??
    join(repoRoot, "vault", "config", "critical-configs.tar.enc");

  logStep("Checking config backup verification prerequisites");

  if (passphrase.length < 16) {
    throw new Error(
      "Set SCRY_CONFIG_BACKUP_PASSPHRASE with at least 16 characters before verification.",
    );
  }

  if (!existsSync(encryptedFile)) {
    throw new Error(`Encrypted backup file not found: ${encryptedFile}`);
  }

  const encrypted = readFileSync(encryptedFile);
  const headerLength =
    CONFIG_FORMAT_MAGIC.length +
    CONFIG_SALT_LENGTH +
    CONFIG_IV_LENGTH +
    CONFIG_AUTH_TAG_LENGTH;

  if (encrypted.length <= headerLength) {
    throw new Error("Encrypted backup payload is too short or malformed.");
  }

  const magic = encrypted
    .subarray(0, CONFIG_FORMAT_MAGIC.length)
    .toString("utf8");
  if (magic !== CONFIG_FORMAT_MAGIC) {
    throw new Error(
      `Unexpected backup format magic: ${magic} (expected ${CONFIG_FORMAT_MAGIC})`,
    );
  }

  const saltStart = CONFIG_FORMAT_MAGIC.length;
  const ivStart = saltStart + CONFIG_SALT_LENGTH;
  const tagStart = ivStart + CONFIG_IV_LENGTH;
  const cipherStart = tagStart + CONFIG_AUTH_TAG_LENGTH;

  const salt = encrypted.subarray(saltStart, ivStart);
  const iv = encrypted.subarray(ivStart, tagStart);
  const authTag = encrypted.subarray(tagStart, cipherStart);
  const ciphertext = encrypted.subarray(cipherStart);

  const key = pbkdf2Sync(
    passphrase,
    salt,
    CONFIG_KDF_ITERATIONS,
    CONFIG_KEY_LENGTH,
    "sha256",
  );

  logStep("Decrypting and extracting backup payload to temp workspace");

  const tempDir = mkdtempSync(join(tmpdir(), "scry-config-verify-"));

  try {
    const decipher = createDecipheriv("aes-256-gcm", key, iv);
    decipher.setAuthTag(authTag);

    let plaintext: Buffer;
    try {
      plaintext = Buffer.concat([
        decipher.update(ciphertext),
        decipher.final(),
      ]);
    } catch {
      throw new Error(
        "Failed to decrypt and authenticate config backup. Check passphrase and artifact integrity.",
      );
    }

    const tarPath = join(tempDir, "critical-configs.tar");
    const extractDir = join(tempDir, "restore-preview");

    writeFileSync(tarPath, plaintext);
    runOrThrow(["tar", "-xf", tarPath, "-C", tempDir]);

    logStep("Validating required restore paths");
    const missing: string[] = [];

    for (const relativePath of REQUIRED_RESTORE_PATHS) {
      const absolutePath = join(tempDir, relativePath);
      if (!existsSync(absolutePath)) {
        missing.push(relativePath);
      }
    }

    if (missing.length > 0) {
      throw new Error(
        `Restore preview missing required paths: ${missing.join(", ")}`,
      );
    }

    const tarSize = statSync(tarPath).size;

    logStep("Config backup verification passed");
    console.log(`artifact: ${encryptedFile}`);
    console.log(`decrypted tar bytes: ${tarSize}`);
    console.log(
      `required restore paths: ${REQUIRED_RESTORE_PATHS.length} present`,
    );
    console.log(`preview root: ${tempDir}`);
    console.log(`home reference: ${process.env.HOME ?? homedir()}`);

    // Keep a deterministic extracted root marker for debugging output readability.
    if (!existsSync(extractDir)) {
      // no-op, extraction is directly into tempDir based on stored relative paths.
    }
  } finally {
    rmSync(tempDir, { recursive: true, force: true });
  }
}

File: scripts/tasks/verify_config_backup.py
"""Verify encrypted config backup can be decrypted and matches recorded metadata."""

from __future__ import annotations

import hashlib
import json
import os
import shutil
import tempfile
from pathlib import Path

from scripts.common import log_step, run_or_throw
from scripts.crypto import CryptoFormat, decrypt

CONFIG_FORMAT = CryptoFormat(magic="SCRYCFG1")


def _load_metadata(metadata_file: Path) -> dict:
    if not metadata_file.exists():
        raise RuntimeError(f"Config backup metadata file not found: {metadata_file}")

    try:
        metadata = json.loads(metadata_file.read_text())
    except json.JSONDecodeError as exc:
        raise RuntimeError(f"Config backup metadata is invalid JSON: {metadata_file}") from exc

    included_paths = metadata.get("includedPaths")
    if not isinstance(included_paths, list) or not all(
        isinstance(item, str) and item for item in included_paths
    ):
        raise RuntimeError(
            "Config backup metadata is missing a valid includedPaths list."
        )

    return metadata


def verify_config_backup() -> None:
    passphrase = os.environ.get("SCRY_CONFIG_BACKUP_PASSPHRASE", "")
    repo_root = Path.cwd().resolve()
    encrypted_file = Path(
        os.environ.get(
            "SCRY_CONFIG_BACKUP_FILE",
            str(repo_root / "vault" / "config" / "critical-configs.tar.enc"),
        )
    )
    metadata_file = Path(
        os.environ.get(
            "SCRY_CONFIG_METADATA_FILE",
            str(repo_root / "vault" / "config" / "critical-configs.meta.json"),
        )
    )

    log_step("Checking config backup verification prerequisites")

    if len(passphrase) < 16:
        raise RuntimeError(
            "Set SCRY_CONFIG_BACKUP_PASSPHRASE with at least 16 characters before verification."
        )

    if not encrypted_file.exists():
        raise RuntimeError(f"Encrypted backup file not found: {encrypted_file}")

    metadata = _load_metadata(metadata_file)
    encrypted = encrypted_file.read_bytes()
    expected_sha256 = metadata.get("encryptedBackupSha256")
    if expected_sha256:
        actual_sha256 = hashlib.sha256(encrypted).hexdigest()
        if actual_sha256 != expected_sha256:
            raise RuntimeError(
                "Encrypted backup SHA-256 does not match metadata. "
                f"metadata={expected_sha256} actual={actual_sha256}"
            )

    log_step("Decrypting and extracting backup payload to temp workspace")

    temp_dir = Path(tempfile.mkdtemp(prefix="scry-config-verify-"))

    try:
        plaintext = decrypt(encrypted, passphrase, CONFIG_FORMAT)
        tar_path = temp_dir / "critical-configs.tar"
        tar_path.write_bytes(plaintext)

        run_or_throw(["tar", "-xf", str(tar_path), "-C", str(temp_dir)])

        log_step("Validating recorded backup contents")
        missing: list[str] = []
        included_paths = metadata["includedPaths"]

        for rel_path in included_paths:
            abs_path = temp_dir / rel_path
            if not abs_path.exists():
                missing.append(rel_path)

        if missing:
            raise RuntimeError(
                f"Restore preview missing recorded paths: {', '.join(missing)}"
            )

        tar_size = tar_path.stat().st_size

        log_step("Config backup verification passed")
        print(f"artifact: {encrypted_file}")
        print(f"metadata: {metadata_file}")
        print(f"decrypted tar bytes: {tar_size}")
        print(f"recorded backup paths: {len(included_paths)} present")
        print(f"source fingerprint: {metadata.get('sourceFingerprint', 'unknown')}")
        print(f"home reference: {os.environ.get('HOME', str(Path.home()))}")
    finally:
        shutil.rmtree(temp_dir, ignore_errors=True)
